### PR TITLE
Show analytics of your points mining

### PR DESCRIPTION
**Pull request by Tkd-Alex**
# Description

You can navigate to localhost:5000 and see a chart with your points history.
We collect every update like watch, claim, prediction, etc.

The following screenshot are just an example.

Screenshot

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md)
- [x] My changes generate no new warnings
- [x] Any dependent changes have been updated in requirements.txt

### CODE REVIEW

**Comment by dmunozv04** [comment]
Could you add variables for the server port and listening address?

**Comment by Tkd-Alex** [comment]
Of course, I'll do :)

**Comment by dmunozv04** [comment]
> Of course, I'll do :)

Thanks!!

**Comment by RakSrinaNa** [comment]
This doesn't work on Windows.

```
Traceback (most recent call last):
  File "D:\Documents\Programming\PyCharm\Twitch-Channel-Points-Miner-v2\main.py", line 33, in <module>
    twitch_miner.analytics(host="127.0.0.1", port=5000)
  File "D:\Documents\Programming\PyCharm\Twitch-Channel-Points-Miner-v2\TwitchChannelPointsMiner\TwitchChannelPointsMiner.py", line 113, in analytics
    http_server.start()
  File "C:\Python39\lib\multiprocessing\process.py", line 121, in start
    self._popen = self._Popen(self)
  File "C:\Python39\lib\multiprocessing\context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "C:\Python39\lib\multiprocessing\context.py", line 327, in _Popen
    return Popen(process_obj)
  File "C:\Python39\lib\multiprocessing\popen_spawn_win32.py", line 93, in __init__
    reduction.dump(process_obj, to_child)
  File "C:\Python39\lib\multiprocessing\reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
TypeError: cannot pickle '_thread.lock' object
```

Googling that error, it doesn't seem that uncommon. I'm don't know much about multithreading and Python, but from what I understood the problem comes from the fact that you build the `app` variable before starting the new thread. This then causes a problem when the new process is created as it can't serialize that variable.
Now this probably doesn't happen on UNIX systems as the way processes are forked is different?

Making that change seemed to make it work for Windows too, but I don't really know what does it imply/changes:
```patch
Index: TwitchChannelPointsMiner/classes/AnalyticsServer.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/TwitchChannelPointsMiner/classes/AnalyticsServer.py b/TwitchChannelPointsMiner/classes/AnalyticsServer.py
--- a/TwitchChannelPointsMiner/classes/AnalyticsServer.py	(revision acb9267c02e675f3df592d11f79882b3cf7a784e)
+++ b/TwitchChannelPointsMiner/classes/AnalyticsServer.py	(date 1614434499050)
@@ -1,5 +1,6 @@
 import logging
 import os
+import threading
 from multiprocessing import Process
 from pathlib import Path
 
@@ -34,7 +35,7 @@
     return render_template("charts.html", streamers=",".join(streamers_available()))
 
 
-class AnalyticsServer(Process):
+class AnalyticsServer(threading.Thread):
     def __init__(self, host="127.0.0.1", port=5000):
         super(AnalyticsServer, self).__init__()
```

**Comment by Tkd-Alex** [comment]
> This doesn't work on Windows.

Before https://github.com/Tkd-Alex/Twitch-Channel-Points-Miner-v2/pull/96/commits/acb9267c02e675f3df592d11f79882b3cf7a784e , works? Or not?

**Comment by RakSrinaNa** [comment]
> > This doesn't work on Windows.
> 
> Before [acb9267](https://github.com/Tkd-Alex/Twitch-Channel-Points-Miner-v2/commit/acb9267c02e675f3df592d11f79882b3cf7a784e) , works? Or not?

ab4b4de064eb20165fbba98da8dde82efa810f31 have the same problem.

**Comment by Tkd-Alex** [comment]
Can you try to move:
```python
        self.app = Flask(
            __name__,
            template_folder=os.path.join(Path().absolute(), "assets"),
            static_folder=os.path.join(Path().absolute(), "assets"),
        )
        self.app.add_url_rule("/", "index", index)
        self.app.add_url_rule("/json/<string:streamer>", "json", read_json)
```

Inside 
```python 
def run(self): 
```

please?

**Comment by Tkd-Alex** [comment]
The use of `run()` directly instead of `start()` seems to be a solution: https://www.programmersought.com/article/98211894530/

**Comment by RakSrinaNa** [comment]
> Can you try to move:
> 
> ```python
>         self.app = Flask(
>             __name__,
>             template_folder=os.path.join(Path().absolute(), "assets"),
>             static_folder=os.path.join(Path().absolute(), "assets"),
>         )
>         self.app.add_url_rule("/", "index", index)
>         self.app.add_url_rule("/json/<string:streamer>", "json", read_json)
> ```
> 
> Inside
> 
> ```python
> def run(self): 
> ```
> 
> please?


It brings a whole lot of new problems.

First one is that it reloads the main script and detects it:
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Python39\lib\multiprocessing\spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "C:\Python39\lib\multiprocessing\spawn.py", line 125, in _main
    prepare(preparation_data)
  File "C:\Python39\lib\multiprocessing\spawn.py", line 236, in prepare
    _fixup_main_from_path(data['init_main_from_path'])
  File "C:\Python39\lib\multiprocessing\spawn.py", line 287, in _fixup_main_from_path
    main_content = runpy.run_path(main_path,
  File "C:\Python39\lib\runpy.py", line 268, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "C:\Python39\lib\runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "C:\Python39\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "D:\Documents\Programming\PyCharm\Twitch-Channel-Points-Miner-v2\main.py", line 32, in <module>
    twitch_miner.analytics(host="127.0.0.1", port=5000)
  File "D:\Documents\Programming\PyCharm\Twitch-Channel-Points-Miner-v2\TwitchChannelPointsMiner\TwitchChannelPointsMiner.py", line 113, in analytics
    http_server.start()
  File "C:\Python39\lib\multiprocessing\process.py", line 121, in start
    self._popen = self._Popen(self)
  File "C:\Python39\lib\multiprocessing\context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "C:\Python39\lib\multiprocessing\context.py", line 327, in _Popen
    return Popen(process_obj)
  File "C:\Python39\lib\multiprocessing\popen_spawn_win32.py", line 45, in __init__
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "C:\Python39\lib\multiprocessing\spawn.py", line 154, in get_preparation_data
    _check_not_importing_main()
  File "C:\Python39\lib\multiprocessing\spawn.py", line 134, in _check_not_importing_main
    raise RuntimeError('''
RuntimeError: 
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
```

Making sure that the code isn't ran twitch before finishing initializing with the `if __name__ == '__main__':` all around the "main" makes the script load fine. However when trying to access the webpage, it leads to this error:
```
[2021-02-27 16:11:57,189] ERROR in app: Exception on / [GET]
Traceback (most recent call last):
  File "D:\Documents\Programming\PyCharm\Twitch-Channel-Points-Miner-v2\venv\lib\site-packages\flask\app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "D:\Documents\Programming\PyCharm\Twitch-Channel-Points-Miner-v2\venv\lib\site-packages\flask\app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "D:\Documents\Programming\PyCharm\Twitch-Channel-Points-Miner-v2\venv\lib\site-packages\flask\app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "D:\Documents\Programming\PyCharm\Twitch-Channel-Points-Miner-v2\venv\lib\site-packages\flask\_compat.py", line 39, in reraise
    raise value
  File "D:\Documents\Programming\PyCharm\Twitch-Channel-Points-Miner-v2\venv\lib\site-packages\flask\app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "D:\Documents\Programming\PyCharm\Twitch-Channel-Points-Miner-v2\venv\lib\site-packages\flask\app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "D:\Documents\Programming\PyCharm\Twitch-Channel-Points-Miner-v2\TwitchChannelPointsMiner\classes\AnalyticsServer.py", line 34, in index
    return render_template("charts.html", streamers=",".join(streamers_available()))
  File "D:\Documents\Programming\PyCharm\Twitch-Channel-Points-Miner-v2\TwitchChannelPointsMiner\classes\AnalyticsServer.py", line 15, in streamers_available
    path = Settings.analytics_path
AttributeError: type object 'Settings' has no attribute 'analytics_path'
```

Probably because the whole creation phase isn't executed in the new process that is created. Something has to be done in the `__init__` of the `AnalyticsServer` to serialize the settings into the new process.

**Comment by RakSrinaNa** [comment]
> The use of `run()` directly instead of `start()` seems to be a solution: https://www.programmersought.com/article/98211894530/

This makes the script hang on the run function as `self.app.run(host=self.host, port=self.port, threaded=True)` seems to be blocking.

**Comment by RakSrinaNa** [comment]
BTW this windows issue may reveal a bigger problem. It seems like the variables aren't shared between your different processes. So if you update the streamer list (which may come with #75) it maybe won't be updated into the new process (as it only got a copy of that object).

https://docs.python.org/3/library/multiprocessing.html#sharing-state-between-processes
https://stackoverflow.com/a/3697322/3281185

**Comment by Tkd-Alex** [comment]
> BTW this windows issue may reveal a bigger problem. It seems like the variables aren't shared between your different processes. So if you update the streamer list (which may come with #75) it maybe won't be updated into the new process (as it only got a copy of that object).
> 
> https://docs.python.org/3/library/multiprocessing.html#sharing-state-between-processes
> https://stackoverflow.com/a/3697322/3281185
> 
> A thread (which fixes the problem) will use the same memory space. While a Process will be completely new process that will have it's own memory etc. So in that case you'll have to do extra work to make the 2 process to communicate between each other (to share data).

I don't need to share the streamers list because I got streamer's username listing .json file in folder.
The only variable shared is: `Settings.analytics_path`
___________
But, If the use of Thread fix all the problems why we should care about it? :rofl:

**Comment by RakSrinaNa** [comment]
> I don't need to share the streamers list because I got streamer's username listing .json file in folder.
> The only variable shared is: Settings.analytics_path

Then I guess it should be fine.

> But, If the use of Thread fix all the problems why we should care about it? 🤣

I don't know if the use of Process was intentional or not. I think Threads are more commonly used (I think) as it's simpler. Processes often comes when performance is a bit more on the line. But for this project I think a thread should do the job just fine.

**Comment by kelukelol** [comment]
kw2wda14
Working perfectly on windows after https://github.com/Tkd-Alex/Twitch-Channel-Points-Miner-v2/pull/96/commits/f6601995623072b05ad45c1293e5dd9294ca3255.

**Comment by RakSrinaNa** [comment]
Some suggestions

---
1. I don't know what is doable with the char lib but if possible it would be nice to have the time that also appears when a point is focused.

As the points can vary a lot through time it'd nice to have that value indicated too. Especially if we want to go check something in the logs to see the detail of what happened.
Of course we can use the axis on the bottom to estimate the time, but it'd be great to have the actual value directly shown.

image

---
2. Make it so that the currently focused streamer get refreshed every X minutes. 

I find it useful to use this graph to track a bit what is happening. When moving between tabs it gets refreshed so everything is up to date. But if you want to stay on a particular streamer and have the graph in the background/side to monitor it, then it'll never refresh.
Maybe refreshing it every 5-10 minutes could be great. It shouldn't add perform too many requests on the server and still keep the graph updated. Or could make that value configurable.

**Comment by RakSrinaNa** [comment]
I think the graph gets a bit buggy when it comes to predictions. I didn't look how this works in the code, but from what I can guess with the graph it only gets updates by events concerning points change (WATCH, WATCH_STREAK, PREDICTION, REFUND, ...).

image

This leads to weird things like the image shown above. What probably happens is that when a bet is placed, the points aren't directly updated (as there's no "prediction_placed" or "prediction_lost" event for points) and thus stay "too high".
Also as there's the interpolation between points making those off the curve.

image

---

Globally, I personally think that the way predictions are handled on the graph are confusing. Currently it is showing wins/losses and nothing else.
To me, as it is plotting points over time, it'd make more sense to have:
* A point when a prediction is made, indicating why we lost points
* A point if we have a win/refund, indicating why we won those points
* (maybe) A point for a loss. Though as a loss is always 0 doesn't makes much sense in the graph and adds a point for "nothing"

Also having `+xxx WIN...` and `+0 LOSE` doesn't make much sense. Could just put `WIN...` and `LOSE...`. We can get the points won/lost by looking at the graph and would reduce the size of the labels that take up quite some space.

Basically something like:
image

I don't know what others think of that?

**Comment by Tkd-Alex** [comment]
> I didn't look how this works in the code, but from what I can guess with the graph it only gets updates by events concerning points change (WATCH, WATCH_STREAK, PREDICTION, REFUND, ...)

https://github.com/Tkd-Alex/Twitch-Channel-Points-Miner-v2/blob/30455e958579d0d42226ec5d6f78aa37bd2f274a/TwitchChannelPointsMiner/classes/WebSocketsPool.py#L180-L183

The updates are saved for each `community-points-user-v1` earned and spent.
As you can see after `points-spent` (followed by `prediction-made`) we have update channel points. 
______
> Globally, I personally think that the way predictions are handled on the graph are confusing

I agree :rofl: 
I want to do the update today, but I can't access to my Raspberry where I have all the data. Hope to do asap

**Comment by RakSrinaNa** [comment]
With the latest commit it already makes more sense:
image

Though for streamers that uses the predictions quite a lot, it makes the graph quite cluttered and hard to read.
I don't know the limitations of the graph lib but it'd be nice if either:
* pointAnnotation can be shown/hidden
* display the reason in the tooltip. eg: 
```
date
-------------
streamer_name: points
reason
```

---

As a side note, to avoid having points being up in the air and reflect a bit more the real values I tried to set the curse type to `stepline` (not in the pic tho). It works rather well. The point annotations are on the line, and reflects better the actual value in our wallet. Though the graph doesn't look as nice :laughing:

**Comment by Tkd-Alex** [comment]
@RakSrinaNa 
It's possible to have this .json ?

**Comment by RakSrinaNa** [comment]
I have several with lots of points.

[Analytics.zip](https://github.com/Tkd-Alex/Twitch-Channel-Points-Miner-v2/files/6071981/Analytics.zip)

**Comment by Tkd-Alex** [comment]
image

What about using vertical annotation for bet and watch streak event? :/

**Comment by Tkd-Alex** [comment]
image


For example here the updated tooltips with 
> Streamer name
> **Points**: 2343245
> **Reason**: Reason 

The datetime it's duplicated in current chart, so I've deleted, and now we have only one dt value.

**Comment by RakSrinaNa** [comment]
Yeah like that it'd be good. You get the overall chart without anything in the way, and if you need more info just hover in the area you wanna know. Plus the tooltip can have whatever info we need, and the size of the text inside won't matter as it's not always shown (useful for predictions).

File: assets/charts.html [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+                <a style="text-decoration: none;" href="https://github.com/Tkd-Alex/Twitch-Channel-Points-Miner-v2">
+                    <img alt="GitHub last commit" src="https://img.shields.io/github/last-commit/Tkd-Alex/Twitch-Channel-Points-Miner-v2" />
+                </a>
+            </div>
+        </div>
+        <br>
+
+        <div class="tabs">
+            <ul>
+                {% for streamer in streamers.split(',') %}
```
Comment (RakSrinaNa) [inline assets/charts.html:44]: Would be nice to have the list of streamers sorted. When having quite a lot of streamers it's hard to find them in a random order.

```suggestion
                {% for streamer in streamers.split(',')|sort %}
```

**Comment by RakSrinaNa** [comment]
Little update with latest commits:
* Auto refresh of the current graph works fine
* The tooltips work fine too. 
Honestly I feel like it kinda makes the annotations a bit useless now; everything that is needed is inside the tooltip (but no need to remove them as they're togglable). Especially if #103 is fixed it should also include the predictions made.
* Points are not off the curve anymore as there isn't points anymore :laughing: . Though the prediction made vertical bar is in between 2 points so we don't really know the impact it had. Solving #103 should make it be on a point and thus make the graph show our actual balance even when points are spent (without having to wait for a points-earned event).

**Comment by RakSrinaNa** [comment]
Fix for #103 is working fine. Graph now updates correctly when points are spent.

---

A little issue with the annotations on the graph:
* Load a chart with annotations
* Disable annotations
* Click the little "home" button on the top right of the chart (also work if zooming or unzooming using the buttons or by selecting an area)

=> Some annotations reappear even though they are disabled

File: assets/charts.html [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+            },
+        },
+        yaxis: {
+            title: {
+                text: 'Channel points'
+            },
+        },
+        xaxis: {
+            type: 'datetime'
+        },
```
Comment (RakSrinaNa) [inline assets/charts.html:114]: The times that are saved into the json file are the local datetime. Though by default the axis will interpret it as being UTC datetimes.

As my server has a timezone of CET (UTC+1), all the dates in the graph are shifted by an hour (13h, becomes 12h).

Either we need to save dates as UTC inside the json (which is in my opinion the best/safest). Or we can tell the axis to load them as local time (by assuming the tz of the browser is the same as the tz of the server):
```suggestion
        xaxis: {
            type: 'datetime',
            labels: {
                datetimeUTC: false
            }
        },
```

File: TwitchChannelPointsMiner/classes/entities/Streamer.py [hunk 3]
```diff
@@ -146,3 +153,44 @@ def drops_condition(self):
...
+            }
+            self.__save_json("annotations", data)
+
+    def persistent_series(self, event_type="Watch"):
+        self.__save_json("series", event_type=event_type)
+
+    def __save_json(self, key, data={}, event_type="Watch"):
+        # https://stackoverflow.com/questions/4676195/why-do-i-need-to-multiply-unix-timestamps-by-1000-in-javascript
+        # data.update({"x": round(time.time() * 1000)})
+        now = datetime.now().replace(microsecond=0)
```
[the remark this answers is not in the document]
Comment (RakSrinaNa) [reply TwitchChannelPointsMiner/classes/entities/Streamer.py:181]: Yeah seem to be a better approach than what I highlighted.
Mine was just to raise the problem, but saving as UTC is way better as tz is known. Then the end user browser will adapt the dates to its own tz.

**Comment by Jarsa132** [comment]
I would like to see night mode in this

**Comment by Tkd-Alex** [comment]
> I would like to see night mode in this

Me too :rofl:

File: TwitchChannelPointsMiner/classes/entities/Streamer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (RakSrinaNa) [reply TwitchChannelPointsMiner/classes/entities/Streamer.py]: Actually 24bf4c4 makes things worse (now the date is 2h behind the real date). I think that python timezone function already transforms it into utc.

image

I did the tests at around 11h20 local time (UTC+1 / so 10h20 UTC).
The first timestamp `1615368052.0` is `Wed Mar 10 2021 09:20:52 GMT+0000`, so off by an hour.
The second timestamp `1615371706.0` is `Wed Mar 10 2021 10:21:46 GMT+0000`, so correct.

Maybe the timestamps were right in the json file, but somehow the chart didn't adapt it correctly to the TZ of the browser.

**Comment by RakSrinaNa** [comment]
Added a comment on the timezone changes (https://github.com/Tkd-Alex/Twitch-Channel-Points-Miner-v2/pull/96#discussion_r591329526).

Other than that the last commits seem to work fine for the annotations.

File: TwitchChannelPointsMiner/classes/entities/Streamer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Tkd-Alex) [reply TwitchChannelPointsMiner/classes/entities/Streamer.py]: uhm, this is not the same problem of: https://github.com/Tkd-Alex/Twitch-Channel-Points-Miner-v2/issues/90 ?

**Comment by Tkd-Alex** [comment]
@Jarsa132 
image

**Comment by Jarsa132** [comment]
> @Jarsa132
> image

When we will be able to use this?

File: TwitchChannelPointsMiner/classes/entities/Streamer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Tkd-Alex) [reply TwitchChannelPointsMiner/classes/entities/Streamer.py:181]: https://stackoverflow.com/questions/64020836/apexcharts-time-not-matching-javascript-python
https://github.com/Tkd-Alex/Twitch-Channel-Points-Miner-v2/pull/96/commits/108702e7a6b4c8f377831a87591f0fcd56fa4a2a

Your previous solution was right.
https://github.com/Tkd-Alex/Twitch-Channel-Points-Miner-v2/pull/96#discussion_r588370197

Tested and working on my PC. 
image

With `datetimeUTC: false` I've the value showed in `Your time zone`

**Comment by Tkd-Alex** [comment]
> When we will be able to use this?

The dark mode It's present in https://github.com/Tkd-Alex/Twitch-Channel-Points-Miner-v2/pull/96/commits/c40ed368e402baa2f4bbd3f3c2a310d6528abc8b
I don't know when we will merge on master-branch but I think very soon.

**Comment by Tkd-Alex** [comment]
Better?
image image

**Comment by Tkd-Alex** [comment]
Another issue for the datetime was reported here: https://github.com/apexcharts/apexcharts.js/issues/110

**Comment by Tkd-Alex** [comment]
image image

**Comment by RakSrinaNa** [comment]
For the texts, yeah looks goo like that, it's readable.

---

For the apexcharts issue, seems like the solution offered here is to add
```js
xaxis: {
    type: "datetime",
    labels: {
      datetimeUTC: false
    }
  },
```

They gave an example with https://codepen.io/drpuur/pen/ExaQWGx . Here the first timestamp is `Sun Jan 05 2020 23:00:00 GMT+0000` and to me is rendered well as `Mon Jan 06 2020 00:00:00 GMT+0100` in the browser.

I guess that this field `datetimeUTC` forces a date to be rendered as UTC if it is `true` (default?), and converts the date into the local TZ if set to `false`.

**Comment by Tkd-Alex** [comment]
https://apexcharts.com/docs/options/xaxis/ 
Yes `datetimeUTC: true` - It's the default value.
> When turned on, local datetime is converted into UTC. Turn it off if you supply date with timezone info and want to preserve it.

The best solution for us, It's to use `datetimeUTC: false` and leave the text showed in https://github.com/Tkd-Alex/Twitch-Channel-Points-Miner-v2/pull/96#issuecomment-795461903

**Comment by mrcsdl** [comment]
do i need to setup a apache web server or something like this? I cant get to the webpanel "ERR_CONNECTION_REFUSED"

Can someone give me a short tip how to set this up?

**Comment by Jarsa132** [comment]
> do i need to setup a apache web server or something like this? I cant get to the webpanel "ERR_CONNECTION_REFUSED"
> 
> Can someone give me a short tip how to set this up?

Any other webserver is not required because this project uses flask.
If you want to connect from one to second machine that have that webpanel you have to use `0.0.0.0` instead of `127.0.0.1`
Make sure you gave correct port in URL (for example localhost:5000)

**Comment by Tkd-Alex** [comment]
Seems to be stable.
I'll merge asap, anyone has founded problems?

**Comment by kelukelol** [comment]
> Seems to be stable.
> I'll merge asap, anyone has founded problems?

I was using it for more than 4 days non-stop, no issues have been found

**Review by Rakambda — APPROVED** [review]

**Comment by Diogovski** [comment]
> > Seems to be stable.
> > I'll merge asap, anyone has founded problems?
> 
> I was using it for more than 4 days non-stop, no issues have been found

you used the bot for 24 hours straight during 4 days or with intervals?

**Comment by kelukelol** [comment]
> > > Seems to be stable.
> > > I'll merge asap, anyone has founded problems?
> > 
> > 
> > I was using it for more than 4 days non-stop, no issues have been found
> 
> you used the bot for 24 hours straight during 4 days or with intervals?

For 24 hours straight. What's the issue?

**Comment by Diogovski** [comment]
No issue at all, i just thought twitch would have some limits where it would maybe idk block your account or something.

**Comment by darkboy245** [comment]
How do you enable this on a Linux VPS?

**Comment by Tkd-Alex** [comment]
> How do you enable this on a Linux VPS?

https://github.com/Tkd-Alex/Twitch-Channel-Points-Miner-v2#analytics

**Comment by darkboy245** [comment]
> 
> 
> > How do you enable this on a Linux VPS?
> 
> https://github.com/Tkd-Alex/Twitch-Channel-Points-Miner-v2#analytics

Yea figured it out, thanks

**Comment by Barmem** [comment]
again, how do i enable it? 
instructions [here](https://github.com/Tkd-Alex/Twitch-Channel-Points-Miner-v2#analytics) are not that clear

**Comment by Glass47** [comment]
image

**Comment by Glass47** [comment]
When i add this code it just gets stuck here, can you tell me where i need to put this code, this is frustrating because I HAVE no one to talk to nor i am knowledgeable enough to google this. This is driving me crazy..

**Comment by ColinShark** [comment]
@Glass47 Please [create a new issue](https://github.com/Tkd-Alex/Twitch-Channel-Points-Miner-v2/issues/new/choose) instead of trying to revive a long merged PR.

**Comment by invahak** [comment]
help) when I added the analytics code, everything went wrong for me.
image
image

**Comment by Hotman75** [comment]
@invahak you already use the TwitchChannelPointsMiner method below the code you added.
If I am correct, you just copied the example, so you use the mine method too.
So what you need to do is just add the line `twitch_miner.analytics(host="127.0.0.1", port=5000, refresh=5, days_ago=7)` between the TwitchChannelPointsMiner and mine methods.

**Comment by tmaarcxs** [comment]
I can`t find how to enable the analysis panel, I have added all the code and configured everything but the log still gives me this outuput:
"[analytics]: Can't start analytics(), please set enable_analytics=True "

I don't know where to enable the flag, please help

**Comment by tmaarcxs** [comment]
Thanks for your response @rdavydov, thanks to what you have told me, now it notifies that the web has been opened in the configuration that I have set, localhost:5000.
But it seems that although everything seems to be correct, when I try to enter with the browser there is no response :(

I attach the dockerfile and the file run.py
// I changed to .txt to allow github to upload it

Thanks in advance.

[run.txt](https://github.com/Tkd-Alex/Twitch-Channel-Points-Miner-v2/files/10078801/run.txt)
[docker-compose.txt](https://github.com/Tkd-Alex/Twitch-Channel-Points-Miner-v2/files/10078802/docker-compose.txt)

**Comment by rdavydov** [comment]
@tmaarcxs https://github.com/rdavydov/Twitch-Channel-Points-Miner-v2/issues/60

**Comment by justejules** [comment]
😣😭😭😭😭😭

25/09/23 21:06:45 - ERROR - [error]: error from callback <function WebSocketsPool.on_message at 0x7f9f8c68b0>: Error while trying to listen for a topic: {'type': 'RESPONSE', 'error': 'ERR_BADAUTH', 'nonce': 'rBaN0W0NMupQXUhf1dMoXRYzeoFe2G'}                                                                                                25/09/23 21:06:45 - ERROR - [on_error]: #0 - WebSocket error: Error while trying to listen for a topic: {'type': 'RESPONSE', 'error': 'ERR_BADAUTH', 'nonce': 'rBaN0W0NMupQXUhf1dMoXRYzeoFe2G'}                                                                                                                                                 25/09/23 21:06:45 - ERROR - [error]: error from callback <function WebSocketsPool.on_message at 0x7f9f8c68b0>: Error while trying to listen for a topic: {'type': 'RESPONSE', 'error': 'ERR_BADAUTH', 'nonce': '9AzIQQBHyUGvyFJzJndWCNB0Zt30hq'}                                                                                                25/09/23 21:06:45 - ERROR - [on_error]: #0 - WebSocket error: Error while trying to listen for a topic: {'type': 'RESPONSE', 'error': 'ERR_BADAUTH', 'nonce': '9AzIQQBHyUGvyFJzJndWCNB0Zt30hq'}

**Comment by blacktownlad** [comment]
can somebody make a video on this for windows setup please and thanks
